feat(create-docusaurus): ask user for preferred language when no language CLI option provided

Rafael-Martins opened the pull request:
## Pre-flight checklist

- [x] I have read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests).

## Motivation

Following the issue #9401, I created a prompt that just shows up when neither `--javascript` or `--typescript` is passed

## Usage screenshot

![image](https://github.com/facebook/docusaurus/assets/27746333/3bb555b3-1026-4241-894f-dcdeb24c87a2)

## Related issues/PRs
#9401

Rafael-Martins committed:
feat(create-docusaurus): added a language prompt when no flag is informed

slorber commented on packages/create-docusaurus/src/index.ts:
```diff
@@ -452,7 +457,33 @@ export default async function init(
+    language = {[selectedLanguage]: true};
+  }
+
+  const source = await getSource(reqTemplate, templates, {
+    ...cliOptions,
+    ...language,
+  });
```
not super fan of this merge between language and cli options. I don't think we should alter cli options

slorber reviewed (changes requested):
Thanks

It's probably working but the code flow is not super clear, I'd appreciate if you refactor it a bit more.

For example `getSource` should probably get a clear `language: "javascript" | "typescript"` attribute instead of overridden cli options

---

Not sure how hard this would be but if there's only a JS template option and no TS variant available for that template, maybe we could avoid the extra prompt because we'll ask an extra useless question

Rafael-Martins committed:
fix: address pr review

extract language logic into cliUtils
consider template to decide if the prompt will show or not

Rafael-Martins commented:
Hey @slorber, I addressed the comments above, let me now if we can improve it even more!

Josh-Cena commented on packages/docusaurus-utils/src/cliUtils.ts:
```diff
@@ -0,0 +1,44 @@
+  typescript?: boolean;
+};
+
+export async function getLanguage(
```
This function is only used in create-docusaurus. Please don't put it in utils.

Rafael-Martins replied:
Hey, I did this extraction because of @slorber commentary telling that this will be reused in another part of the code, if it is true, I can maybe pass the prompt function as a parameter and remove the dependency, what do you think about it? I can also bring it back to create-docusaurus!

slorber replied:
yes, considering we also want this for swizzle it's fine to add it to docusaurus utils

see also related pr: https://github.com/facebook/docusaurus/pull/9681

slorber commented on packages/docusaurus-utils/src/cliUtils.ts:
```diff
@@ -0,0 +1,44 @@
+  if (languages.typescript || languages.javascript) {
+    return languages;
+  }
+  if (noTsVersionAvailable) {
+    return {javascript: true};
+  }
```
I'm not sure it's a good thing to have this thing here, will try to see where to move it

slorber committed:
Merge branch 'main' into feature-language-selector-create-docussaurus

slorber committed:
Better handling of create-docusaurus javascript/typescript option + refactor

slorber committed:
make sure generate examples is not prompted for js/ts language question

slorber committed:
fix e2e test script

slorber committed:
lint

slorber commented:
Thanks 👍 

I took the opportunity to refactor a bit the CLI because it was quite messy 🤪 

Also adjusting the way we prompt for the language choice because we are going to share some code with another related PR: https://github.com/facebook/docusaurus/pull/9681

will be in v3.2
